### DISCUSSION TITLE
Releasing connections

### DISCUSSION THREAD

**Discussion by JohnCastle4** [ngageoint/geopackage-ios#55] [Q&A] [answered]
We have developed a Cordova plugin for our mobile application and we are using the iOS library, we're also using the Android library, but the issue we found was on iOS. While investigation a performance issue, we found that GPKGConnection.queryWithDistinct gets a resultConnection, but doesn't release it. This resulted in the usedConnections continuously growing.
We're not able to push the changes back for company policy reasons, but we found releasing the Connection to be highly beneficial.
Thought I would raise it here in case there are other changes beyond the one that we are making address the specific issue we have found.

**Comment by bosborn** [comment, marked as answer]
Hello.

Yes, a GPKGResultSet returned from any method should/must be [closed](https://github.com/ngageoint/geopackage-ios/blob/6d738860f7b616f55474c4eeffa946cda3e9172e/geopackage-ios/db/GPKGResultSet.h#L106) immediately when the results are no longer needed.  A SQLite (read result) connection is pulled from the connection pool for the query and not released until the GPKGResultSet is closed.  At the same time, that connection must remain open temporarily to access those results.  Any method in this library that internally opens but does not return the result set, closes that connection before return (unless you found a bug where that is not the case). Any call to this library that is returned any type of Result Set becomes responsible for closing the results as part of normal and expected usage.

Unless I am misunderstanding this doesn't appear to be an issue within the library and upon closing all your query results there was no issue?

Edit: FYI on Android, any type of Cursor or CursorResults should also be closed to avoid issues.

**Reply by JohnCastle4** [reply]
Ah, perfect. We were missing this step of closing the result set after querying. I've changed to fix our code properly now I've understood the correct patterns and that is correctly stopping the usedConenctions from growing. Got a second issue where a sub-select is missing the table, but haven't investigated the cause of that yet.

**Reply by JohnCastle4** [reply]
Still scratching my head with the missing table name. We're calling RTreeIndexTableDao.queryFeatures(false, boundingBox) and SQL like this is generated:
select .... from "table_name" where ... IN (SELECT ... from  where ...)
Looks like the feature table name is populated, but the rtree index table name isn't. Very strange, seems to work fine for a while and then throw this with a random table.
The configuration we are using has quite a lot of tables spanning 3 gpkg files.
I'm definitely missing something, so far I can only catch the error, not where it's going wrong, which appears to be in queryIdsSQLWhere when the bounding box is turned into the full subselect from the rtree table.
Not 100% sure, but I'm pretty sure it's queried this table many times before it falls over. On the basis that I've zoomed in and was panning on the map to trigger the error.
Our handling may be bad because it doesn't open any tables at all after this error.

**Reply by bosborn** [reply]
I'm not entirely sure just from that.  When you create the RTree, there are a set of 4 tables created in the database.
```
rtree_<table_name>_geometry
rtree_<table_name>_geometry_node
rtree_<table_name>_geometry_parent
rtree_<table_name>_geometry_rowid
```

* For using RTree directly, [this test](https://github.com/ngageoint/geopackage-ios/blob/6d738860f7b616f55474c4eeffa946cda3e9172e/geopackage-iosTests/extension/rtree/GPKGRTreeIndexExtensionUtils.m#L17) might help.

* A GeoPackage creation example of [creating RTrees](https://github.com/ngageoint/geopackage-ios/blob/6d738860f7b616f55474c4eeffa946cda3e9172e/geopackage-iosTests/GPKGGeoPackageExample.m#L1396).

* There is also a [FeatureIndexManager](https://github.com/ngageoint/geopackage-ios/blob/master/geopackage-ios/features/index/GPKGFeatureIndexManager.h) with combined logic and querying from the different types of indices. The implementation of which can also be useful as guidance to using RTree directly.
  * [README example](https://github.com/ngageoint/geopackage-ios#objective-c-example) (see "// Index Features")
  * [GeoPackage creation example](https://github.com/ngageoint/geopackage-ios/blob/6d738860f7b616f55474c4eeffa946cda3e9172e/geopackage-iosTests/GPKGGeoPackageExample.m#L842)
  * [Tests](https://github.com/ngageoint/geopackage-ios/blob/master/geopackage-iosTests/features/index/GPKGFeatureIndexManagerUtils.m)

**Reply by JohnCastle4** [reply]
With the help of some console logging I know a bit more now. Configuration has a total of 22 feature layers, but not all of them are visible at every zoom level, when zoomed in, there are a lot of layers visible including text and polygon feature layers.

At some point after panning for a while the app stops with the error above. Tracing back to the code in our plugin the GPKGTreeIndexTableDao returned using tableDaoWithFeatureDao doesn't have a table name, the table._contents property is null and hence the tableName can't be retrieved. With the help of the logging added, I can see this table has been queried many times, since it's the water layer I've seen the rivers and streams rendered on the map many times before this crash.

I suspect this is a symptom of some other issue, could be data related or could be that our calling code has another bug, like a slow leak. Still investigating further, will check your examples as well, we're querying on the bounding box, so it's the base geometry table we're querying. Thank you, for your help.

**Reply by bosborn** [reply]
If it helps, we also use [GPKGGeoPackageCache](https://github.com/ngageoint/geopackage-ios/blob/master/geopackage-ios/GPKGGeoPackageCache.h) to manage multiple open GeoPackages.

Here are our iOS apps that use this library that can be referenced as examples
* [MapCache](https://github.com/ngageoint/geopackage-mapcache-ios) [[manager](https://github.com/search?q=repo%3Angageoint%2Fgeopackage-mapcache-ios+GPKGGeoPackageManager&type=code), [cache](https://github.com/search?q=repo%3Angageoint%2Fgeopackage-mapcache-ios+GPKGGeoPackageCache&type=code), [index](https://github.com/search?q=repo%3Angageoint%2Fgeopackage-mapcache-ios+GPKGFeatureIndexManager&type=code)]
* [MAGE](https://github.com/ngageoint/mage-ios) [[manager](https://github.com/search?q=repo%3Angageoint%2Fmage-ios+GPKGGeoPackageManager&type=code), [cache](https://github.com/search?q=repo%3Angageoint%2Fmage-ios+GPKGGeoPackageCache&type=code), [index](https://github.com/search?q=repo%3Angageoint%2Fmage-ios+GPKGFeatureIndexManager&type=code)]
* [Marlin](https://github.com/ngageoint/marlin-ios) [[manager](https://github.com/search?q=repo%3Angageoint%2Fmarlin-ios+GPKGGeoPackageManager&type=code), [cache](https://github.com/search?q=repo%3Angageoint%2Fmarlin-ios+GPKGGeoPackageCache&type=code), [index](https://github.com/search?q=repo%3Angageoint%2Fmarlin-ios+GPKGFeatureIndexManager&type=code), [rtree](https://github.com/search?q=repo%3Angageoint%2Fmarlin-ios%20rtree&type=code)]

**Reply by JohnCastle4** [reply]
Thanks, we are already GeoPackageCache

**Reply by JohnCastle4** [reply]
It's not data related, the section of map being retrieved isn't related to the crash. It looks like memory, the memory usage just keeps going up. I don't think it's our open layers implementation, this was pretty stable with SpatiaLite for many years. However, we've been forced to migrate away from that library because it's no longer being maintained and the world moves on. Suspect it's our code. We're doing a queryFeatures andBoundingBox and we are now closing the resultset. We pass in the map extents and convert those into a bounding box, then using PROJTransform and SFPGeometryTransform to convert that into a projected bounding box for the layer and that's our bounding box. It seems to work, but it appears not be releasing everything. It also could be our Cordova interface returning an array of the feature ids and geometry that's not handling memory well.

**Reply by bosborn** [reply]
If you are adding features on the map, we've had to remove features no longer visible in map view for performance.  We prefer to use feature tile overlays (dynamically drawn tiles from features). See `// Feature Tile Overlay (dynamically draw tiles from features)` on the [README example](https://github.com/ngageoint/geopackage-ios#objective-c-example).  Queries are still done from map "clicks".  In both cases, we also simplify geometries to reduce the number of points in the displayed or drawn geometry.

**Reply by JohnCastle4** [reply]
Finally figured it out. Android was using the same GeoPackageManager instance as much as possible. Whereas iOS was getting a new one every time and opening a connection. I'm not very familiar with debugging in xCode, but once I found the memory management tools in the debugging the connection pool leak became obvious. Thank you for your help, much appreciated. The app is now much faster and doesn't leak. This discussion can be closed now.
